### PR TITLE
Weight & Bias initializers for Linear, Conv, RNN

### PR DESCRIPTION
**Pull request by kevinzakka**
As mentioned in #5370, here's what adding weight and bias string args to some of the layers could look like.
```python
# use a string
x = nn.Linear(784, 512, weight_init='xavier_uniform', bias_init='zeros')

# use a function
x = nn.Linear(784, 512, weight_init=init.uniform(-5, 5))
x = nn.Linear(784, 512, weight_init=init.xavier_uniform(gain=init.calculate_gain('tanh')))
```

Example of backwards-compatibility:
```python
w = torch.randn(2, 2)
nn.init.constant(w, 1) # prints [[1, 1], [1, 1]]
nn.init.Constant(0)(w) # prints [[0, 0], [0, 0]]
```

Finally, test to show that new implementation of `kaiming` and `xavier` work:
```python
# same size input and output -> in = out = average -> approx same std
W = torch.randn(700, 700)
xavier = nn.init.xavier_uniform(W.clone(), gain=nn.init.calculate_gain('relu'))
kaiming = nn.init.kaiming_uniform(W.clone())
print(torch.std(xavier)) # prints 0.053412361240471495
print(torch.std(kaiming)) # prints 0.05350749617735303

# diff size input and output -> but fan set to avg -> approx same std
W = torch.randn(700, 500)
xavier = nn.init.xavier_uniform(W.clone(), gain=nn.init.calculate_gain('relu'))
kaiming = nn.init.kaiming_uniform(W.clone(), mode="fan_avg")
print(torch.std(xavier)) # prints 0.057737772865026314
print(torch.std(kaiming)) # prints 0.057677882183886975
```

Also added `selu` support to `calculate_gain`.

### CODE REVIEW

**Comment by kevinzakka** [comment]
@apaszke this commit should add backwards compatibility. I don't know if there's a nicer way of writing my base class `Initializer`, so it probably needs checking. 

Also, how would you want to write the documentation for each class? It's not a function anymore so the examples have changed.

Cheers

**Comment by ezyang** [comment]
Sorry we've let this PR sit for a while. What's the testing plan?

**Comment by kevinzakka** [comment]
@ezyang hey Edward, I have a small script I wrote to test out the different initialization strategies but it doesn't conform to pytorch unit testing. Should I go add some functions to `test_nn.py`?

If so, should there be a test for each initialization and for each init strategy (string or function)?

**Comment by apaszke** [comment]
@kevinzakka yes, we'll need tests in `test_nn.py`. Also, can you please remove the string-based option? I don't think it provides any value upon the alternatives

**Comment by kevinzakka** [comment]
@apaszke Deal, going through exam period right now, will do so when I free up :)

**Comment by kevinzakka** [comment]
@apaszke @ezyang I've noticed you guys have added an underscore after the name and introduced deprecation. Should I add that to my PR? Also, why was that change done?

**Comment by ikostrikov** [comment]
This PR can be extremely useful. Let me know if I can help.

**Comment by ezyang** [comment]
@kevinzakka I'm not sure which you are referring to? But yes, the merge conflicts will need to be resolved. Shout if you need help.

**Comment by kevinzakka** [comment]
@ezyang should be done now. There are some conflicts because I used classes and gave them a lower case alias. The lower case alias also uses the deprecation function introduced in the most recent change to this file.

File: test/test_nn.py [hunk 1]
```diff
@@ -4164,7 +4164,7 @@ def test_conv_double_backward_cuda(self):
                                 "\ninp_size: " + str(inp_size) +
                                 "\ndilation: " + str(dilation))
 
-
+# hey kevin
```
Comment (apaszke) [inline test/test_nn.py]: Hello! Can you please remove this comment? 😄

File: torch/nn/init.py [hunk 2]
```diff
@@ -1,239 +1,394 @@
...
+    """
+    Base class for all initializations.
+    """
+    def __new__(cls, *non_tensor_args, **non_tensor_kwargs):
+        first = None
+        if len(non_tensor_args) > 0:
+            first = non_tensor_args[0]
+        if first is None:
+            return object.__new__(cls)
+        if isinstance(first, (torch.Tensor, torch.autograd.Variable)):
```
Comment (apaszke) [inline torch/nn/init.py]: Can you please remove the `Variable` part? It's not necessary anymore

File: torch/nn/init.py [hunk 3]
```diff
@@ -1,239 +1,394 @@
...
+            return cls(*non_tensor_args[1:], **non_tensor_kwargs)(first)
```
Comment (apaszke) [inline torch/nn/init.py]: Isn't this going to enter an infinite loop? Don't you need to use `super` here?

File: torch/nn/init.py [hunk 4]
```diff
@@ -1,239 +1,394 @@
...
+class Initializer(object):
+    """
+    Base class for all initializations.
+    """
+    def __new__(cls, *non_tensor_args, **non_tensor_kwargs):
+        first = None
+        if len(non_tensor_args) > 0:
+            first = non_tensor_args[0]
+        if first is None:
+            return object.__new__(cls)
```
Comment (apaszke) [inline torch/nn/init.py]: This ignores kwargs.

File: torch/nn/init.py [hunk 5]
```diff
@@ -1,239 +1,394 @@
...
+        if isinstance(first, (torch.Tensor, torch.autograd.Variable)):
+            return cls(*non_tensor_args[1:], **non_tensor_kwargs)(first)
+        else:
+            return object.__new__(cls)
```
Comment (apaszke) [inline torch/nn/init.py]: Kwargs ignored again

File: torch/nn/init.py [hunk 6]
```diff
@@ -1,239 +1,394 @@
...
-        raise ValueError("Unsupported nonlinearity {}".format(nonlinearity))
+    def __call__(self, tensor):
+        with torch.no_grad():
+            return tensor.fill_(1)
 
 
-def uniform(tensor, a=0, b=1):
-    """Fills the input Tensor or Variable with values drawn from the uniform
-    distribution :math:`U(a, b)`.
+class Zeros(Initializer):
```
Comment (apaszke) [inline torch/nn/init.py]: Can you please reorder those in the same way as they appeared in the original file? Right now it's hard to review them, because the diffs are really messed up

File: torch/nn/init.py [hunk 7]
```diff
@@ -1,239 +1,394 @@
...
+            for col_idx in range(cols):
+                row_indices = list(range(rows))
+                random.shuffle(row_indices)
+                zero_indices = row_indices[:num_zeros]
+                for row_idx in zero_indices:
+                    tensor[row_idx, col_idx] = 0
+        return tensor
+
+
+class VarianceScaling(object):
```
Comment (apaszke) [inline torch/nn/init.py]: What's this? Any docs?

File: torch/nn/init.py [hunk 8]
```diff
@@ -273,85 +435,140 @@ def kaiming_normal(tensor, a=0, mode='fan_in'):
...
+        >>> he_init(w)
     """
-    fan = _calculate_correct_fan(tensor, mode)
-    gain = calculate_gain('leaky_relu', a)
-    std = gain / math.sqrt(fan)
-    with torch.no_grad():
-        return tensor.normal_(0, std)
+    def __init__(self, a=0, mode="fan_in"):
+        gain = calculate_gain('leaky_relu', a) ** 2
+        self.vs = VarianceScaling(
```
Comment (apaszke) [inline torch/nn/init.py]: It seems that you could make `VarianceScaling` an `Initializer` subclass, and then subclass from it in here. Then, in `__init__` you'd take those arguments, modify them, and pass them to `super`. Then, you don't even need to implement `__call__`

File: torch/nn/modules/batchnorm.py [hunk 9]
```diff
@@ -35,8 +36,8 @@ def reset_parameters(self):
             self.running_mean.zero_()
             self.running_var.fill_(1)
         if self.affine:
-            self.weight.data.uniform_()
-            self.bias.data.zero_()
+            init.Ones(self.weight)
+            init.Zeros(self.bias)
```
Comment (apaszke) [inline torch/nn/modules/batchnorm.py]: 1. This changes the default initialization
2. Why aren't you using the public API?

File: torch/nn/modules/conv.py [hunk 10]
```diff
@@ -35,16 +37,14 @@ def __init__(self, in_channels, out_channels, kernel_size, stride,
...
+        self.bias_init = bias_init
         self.reset_parameters()
 
     def reset_parameters(self):
-        n = self.in_channels
-        for k in self.kernel_size:
-            n *= k
-        stdv = 1. / math.sqrt(n)
-        self.weight.data.uniform_(-stdv, stdv)
+        self.weight_init(self.weight.data)
```
Comment (apaszke) [inline torch/nn/modules/conv.py]: We'll need to set `weight_init` in `__setstate__`.

**Review by apaszke — CHANGES_REQUESTED** [review]
You checked in a few files accidentally. Also needs to have the conflicts resolved.

File: torch/nn/init.py [hunk 11]
```diff
@@ -1,239 +1,394 @@
...
+        num_zeros = int(math.ceil(rows * self.sparsity))
+
+        with torch.no_grad():
+            tensor.normal_(0, self.std)
+            for col_idx in range(cols):
+                row_indices = list(range(rows))
+                random.shuffle(row_indices)
+                zero_indices = row_indices[:num_zeros]
+                for row_idx in zero_indices:
+                    tensor[row_idx, col_idx] = 0
```
Comment (fmassa) [inline torch/nn/init.py]: Could you rebase on top of master? A PR that optimized this operation has been recently merged

**Review by fmassa — CHANGES_REQUESTED** [review]

**Comment by kevinzakka** [comment]
Hey @ezyang @apaszke @fmassa, haven't been very active on the pytorch side of things, but I have some free time and was wondering if I could still work on this PR and submit it. Would appreciate any comments.

**Comment by npuichigo** [comment]
@kevinzakka Any update? It's useful changes
